[MRG] Recurrence quantification analysis (RQA)

bmcfee opened the pull request:
#### Reference Issue
Implements #837 
[added by @lostanlen: closes #837]


#### What does this implement/fix? Explain your changes.

This PR adds RQA methods as described in the reference issue.


#### Any other comments?

I think this is ready for CR.

bmcfee commented:
For completeness, here's how the docstring examples render:

(No gaps or knight moves)
![image](https://user-images.githubusercontent.com/1190540/60387875-dd886500-9a77-11e9-9b6b-4e1aeb9073d9.png)

(Gaps and knight moves allowed)
![image](https://user-images.githubusercontent.com/1190540/60387876-e0835580-9a77-11e9-9128-ad4bfd71c7a7.png)

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+
+    This function implements different forms of RQA as described by
+    Serra, Serra, and Andrzejak [1]_.  These methods take as input
+    a recurrence or cross-similarity matrix `sim`, and calculate the value
```
~what is the difference between a recurrence matrix and a cross-similarity matrix?~
also note that, because "recurrence" is a noun, it might not be obvious to non-native English speakers that it is adjectived and that the noun "matrix" applied to both "recurrence" or "cross-similarity".

lostanlen replied:
bis: i would be in favor of replacing recurrence by self-similarity

bmcfee replied:
I went with recurrence to match the function name, but we have a see-also, so i'm not too picky about it.

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+        - `score[i, j] = 0` otherwise.
+
+    The second method, denoted as `S` [1]_ (equation 4), is similar to the first,
+    but allows for "knight moves" in addition to strict diagonal moves:
```
would be good to be explicit about the reference to the chess game here.

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+    allowing gaps in the alignment that incur some cost, rather than a hard
+    reset to 0 whenever `sim[i, j] == 0`.
+    Gaps are penalized by two additional parameters, `gap_onset` and `gap_extend`,
+    which are deducted from the value of the alignment path every time a gap
```
deducted -> subtracted

(i know they are synonyms, but deducted is often confused for deduced by non-native En speakers)

bmcfee replied:
good catch

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+
+
+    .. [1] Serra, Joan, Xavier Serra, and Ralph G. Andrzejak.
+        "Cross recurrence quantification for cover song identification."
```
Cross-recurrence

bmcfee replied:
The paper title does not have a hyphen.

lostanlen replied:
🤷‍♂

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+    method, and disabling knight moves recovers the first.
+
+
+    .. [1] Serra, Joan, Xavier Serra, and Ralph G. Andrzejak.
```
technically it's Joan Serrà and Xavier Serra, but i can see why you'd want to asciify Joan's last name into Serra

bmcfee replied:
We do set utf-8 encoding on all source files, so there's not really a problem here.  It may cause some difficulty in docstring rendering in REPL or sphinx, but I'll look into it.

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+
+    .. [1] Serra, Joan, Xavier Serra, and Ralph G. Andrzejak.
+        "Cross recurrence quantification for cover song identification."
+        New Journal of Physics 11, no. 9 (2009): 093017.
```
i don't know what this "093017" means. Apparently it corresponds to the `pages` field in BibTeX ... ?

bmcfee replied:
I guess so -- this is from the citation export, so I'm inclined to keep it.

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+    ----------
+    sim : np.ndarray [shape=(N, M), non-negative]
+        The similarity matrix to use as input.
+
```
unnecessary whitespace

bmcfee replied:
where?

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+    knight_moves : bool
+        If `True` (default), allow for "knight moves" in the alignment,
+        e.g., `(n, m) => (n + 1, m + 2)` or `(n + 2, m + 1)`.
+
```
unnecessary whitespace

bmcfee replied:
where?

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+
+    backtrack : bool
+        If `True`, return the alignment path.
+
```
unnecessary whitespace

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+    path : np.ndarray [shape=(k, 2)] (optional)
+        If `backtrack=True`, `path` contains a list of pairs of aligned frames
+        in the best alignment sequence.
+
```
unnecessary whitespace

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+
+    # These are place-holder arrays to limit the points being considered
+    # at each step of the DP
+    #
```
unnecessary whitespace (and again later in this comment block)

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+    # backtracking initialization: the first row and column are all resets
+    # if there's a positive link here, it's an inclusive reset
+    for i in range(sim.shape[0]):
+        if sim[i, 0]:
```
⚠️ ⚠️ ⚠️ 
why is it `if sim[i,0]` and not `if sim[i,0]>0`? can `sim` contain `None` or `False`?

bmcfee replied:
sim can't contain none.  It could contain `False` (dtyped to np.bool); in general, nonzero is truthy and zero is not truthy.

lostanlen replied:
ok i understand now!

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+            break
+
+        # Otherwise, prepend this index and continue
+        idx = [idx[_] + offsets[bt_index][_] for _ in range(len(idx))]
```
is there a reason why you're iterating with the underscore symbol rather than with a letter?

bmcfee replied:
code checker does not like single-letter variables; `_` for this purpose is pythonic though.

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+        # Otherwise, prepend this index and continue
+        idx = [idx[_] + offsets[bt_index][_] for _ in range(len(idx))]
+
+    if not path:
```
can you put a comment describing when this case typically happens?

lostanlen commented on librosa/sequence.py:
```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+    >>> librosa.display.specshow(L_score, x_axis='frames', y_axis='frames')
+    >>> plt.plot(L_path[:, 1], L_path[:, 0], label='Optimal path', color='c')
+    >>> plt.legend()
+    >>> plt.show()
```
can you make sure that the x-axis and y-axis have equal length? ("square" mode for specshow)

bmcfee replied:
specshow doesn't directly support aspect ratios, but i can change the figure size.

bmcfee commented:
Thanks for the CR @lostanlen !  I've addressed most of the issues, and strengthened the tests a little.  Let me know if there's anything else that you think needs changing.

lostanlen commented:
LGTM now! sorry for the false alarms in CR
